File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def solution(enroll, referral, seller, amount):
    tree = {}
    benefit = {}
    tree["minho"] = None
    benefit["minho"] = 0
    
    for name in enroll:
        tree[name] = ""
        benefit[name] = 0
    
    for i in range(len(enroll)):
        cur_member = enroll[i]
        inviter = referral[i]
        
        if inviter == "-":
            tree[cur_member] = "minho"
        else:            
            tree[cur_member] = inviter
        
    for i in range(len(seller)):
        cur_seller = seller[i]
        cur_benefit = amount[i] * 100
        
        inviters_benefit = cur_benefit // 10
        # print("1  :", cur_benefit, inviters_benefit)
        if inviters_benefit < 1:
            benefit[cur_seller] += cur_benefit
            continue
        benefit[cur_seller] += cur_benefit - inviters_benefit
        
        next_taker = tree[cur_seller]
        while next_taker is not None:           
            next_inviters_benefit = inviters_benefit // 10
            # print("fee : ", inviters_benefit, next_inviters_benefit)
            if next_inviters_benefit < 1:
                benefit[next_taker] += inviters_benefit
                break
            benefit[next_taker] += inviters_benefit - next_inviters_benefit
            inviters_benefit = next_inviters_benefit
            next_taker = tree[next_taker]
            
        
    return [benefit[name] for name in enroll]

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    HashMap<String, String> memberReferenceMap = new HashMap<>();
    HashMap<String, Integer> memberSalesMap = new HashMap<>();
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        for(int i = 0; i < enroll.length; i++){
            memberReferenceMap.put(enroll[i], referral[i]);
            memberSalesMap.put(enroll[i], 0);
        }
        
        for(int i = 0; i < seller.length; i++){
            String curSeller = seller[i];
            int curSales = amount[i] * 100;

            // 추천인이 없고 매출도 없으면 종료
            while(!curSeller.equals("-") && curSales > 0){
                String curReference = memberReferenceMap.get(curSeller);
                int curFee = curSales / 10;
                
                // 수수료가 0원이면 트리 상향 중단
                if(curFee == 0){
                    memberSalesMap.put(curSeller, memberSalesMap.get(curSeller) + curSales);
                    break;
                }
                    
                // 수수료가 있으므로 차감 후 판매액 지급
                memberSalesMap.put(curSeller, memberSalesMap.get(curSeller) + curSales - curFee);                                                   
                                
                curSeller = curReference;
                curSales = curFee;                
            }                                    
        }
        
        int[] answer = new int[enroll.length];
        
        for(int i = 0; i < enroll.length; i++){                
            answer[i] = memberSalesMap.get(enroll[i]);
        }            
        
        return answer;
    }
    
    public int getFee(int sales){
        return sales / 10;
    }
}
